index.rst

```diff
@@ -1,29 +1,988 @@
+This requires repeating some of the configuration data model, but it means that settings can be rearranged, canonicalized, and merged with secrets to create a more coherent internal configuration data structure.
+
+.. _models:
+
+Models
+======
+
+FastAPI relies on Pydantic_ for validation and parsing, so all models used by handlers must be Pydantic models.
+This includes the models for form submission as well as JSON POST bodies, when form submission has to be supported.
+It also includes anything returned by a handler in a response body, including error responses.
+
+.. _Pydantic: https://pydantic-docs.helpmanual.io/
+
+Pydantic models
+---------------
+
+Since the Pydantic models are used to generate the API documentation, fields in models should always use the ``Field`` constructor and include as much information as possible about that field.
+For example:
+
+.. code-block:: python
+
+    name: str = Field(
+        ...,
+        title="Name of the group",
+        example="g_special_users",
+        min_length=1,
+        regex=GROUPNAME_REGEX,
+    )
+
+As shown in this example, make as much use as possible of the built-in validation support in Pydantic so that Pydantic plus FastAPI will do basic validity checks on any user input.
+
+``title`` must always be set to a short description of the field (no period at the end).
+``example`` should normally be set.
+If there is a need for longer discussion than will fit in the few words available in ``title``, add ``description``, which can be multiple regular sentences and can even use Markdown formatting if needed.
+For example:
+
+.. code-block:: python
+
+    token_type: TokenType = Field(
+        ...,
+        description=(
+            "Class of token, chosen from:\n\n"
+            "* `session`: An interactive user web session\n"
+            "* `user`: A user-generated token that may be used"
+            " programmatically\n"
+            "* `notebook`: The token delegated to a Jupyter notebook for"
+            " the user\n"
+            "* `internal`: A service-to-service token used for internal"
+            " sub-calls made as part of processing a user request\n"
+            "* `service`: A service-to-service token used for internal calls"
+            " initiated by services, unrelated to a user request\n"
+        ),
+        title="Token type",
+        example="session",
+    )
+
+As you can see from that example, while FastAPI tries to produce good documentation from enums, it's often not clear enough and one may need to hand-craft a good description.
+
+Any field in a model that takes a limited set of values should be defined as a type inheriting from ``Enum``.
+I generally do not make the class also inherit from ``str`` and instead explicitly add ``.value`` to get the string value of an enum.
```

Comment:
Can Pydantic serialize to JSON an Enum that doesn't inherit from `str`, or do you need to configure the serialization? I think this has been the main reason I've inherited from both str and Enum in my enums.

Reply:
Yes, despite all the StackOverflow questions that imply it can't, this does work in Pydantic in most cases. There are a few cases that it can't handle (serializing dictionaries where the enum is the key, for instance), but for simple data members that are enum types, Pydantic apparently adds a custom JSON encoder by default that uses the enum value. I'm using this in Gafaelfawr for `TokenType` and never noticed that I didn't make it inherit from `str`.